dask/dataframe/shuffle.py

```diff
@@ -24,6 +25,196 @@
+        List of required output-partition indices.
+    """
+
+    def __init__(
+        self,
+        name,
+        column,
+        inputs,
+        stage,
+        npartitions,
+        n,
+        k,
+        ignore_index,
+        dep_name,
+        dep_meta,
+        parts_out=None,
+    ):
+        self.column = column
+        self.name = name
+        self.inputs = inputs
+        self.stage = stage
+        self.npartitions = npartitions
+        self.n = n
+        self.k = k
+        self.ignore_index = ignore_index
+        self._name = dep_name
+        self._meta = dep_meta
+        self.parts_out = parts_out or range(len(inputs))
+
+    def __repr__(self):
+        return "ShuffleStage<name='{}', stage={}, k={}, npartitions={}>".format(
+            self.name, self.stage, self.k, self.npartitions
+        )
+
+    @property
+    def _dict(self):
+        """Materialize full dict representation"""
+        if hasattr(self, "_cached_dict"):
+            return self._cached_dict["dsk"]
+        else:
+            dsk = self._rbc_tasks_stage()
+            self._cached_dict = {"dsk": dsk}
+        return self._cached_dict["dsk"]
+
+    def __getitem__(self, key):
+        return self._dict[key]
+
+    def __iter__(self):
+        return iter(self._dict)
+
+    def __len__(self):
+        return len(self._dict)
+
+    def _keys_to_parts(self, keys):
+        parts = set()
+        for key in keys:
+            try:
+                _name, _part = key
+            except ValueError:
+                continue
+            if _name != self.name:
+                continue
+            parts.add(_part)
+        return parts
+
+    def _cull_dependencies(self, keys):
+        """Determine the necessary dependencies to produce `keys`.
+        Does not require graph materialization.
+        """
+        deps = defaultdict(set)
+        parts_out = self._keys_to_parts(keys)
+        inp_part_map = {inp: i for i, inp in enumerate(self.inputs)}
+        for part in parts_out:
+            out = self.inputs[part]
+            for k in range(self.k):
+                _part = inp_part_map[insert(out, self.stage, k)]
+                deps[(self.name, part)].add((self._name, _part))
+        return deps
+
+    def cull(self, keys, all_keys):
+        """Cull the ShuffleStage layer.
+
+        The underlying graph will only include the necessary
+        tasks to produce the keys (indicies) included in `parts_out`.
+        Therefore, "culling" the layer only required us to reset this
+        parameter.
+        """
+        culled_layer = ShuffleStage(
+            self.name,
+            self.column,
+            self.inputs,
+            self.stage,
+            self.npartitions,
+            self.n,
+            self.k,
+            self.ignore_index,
+            self._name,
+            self._meta,
+            parts_out=self._keys_to_parts(keys),
+        )
+        culled_deps = self._cull_dependencies(keys)
```

Comment:
I see that `_cull_dependencies` calls `_keys_to_parts`. If that's expensive, then `cull_dependencies` could take an optional `parts_out` argument. Then we compute it once here and pass it to `cull_dependencies`.

Reply:
Good suggestion - I'll make the change.